pkg/component/runtime/manager_test.go

```diff
@@ -1162,14 +1175,312 @@ LOOP:
+									"type":    "fake",
+									"state":   int(client.UnitStateHealthy),
+									"message": "Fake Healthy",
+									"kill":    rp[1],
+								})
+								err := m.Update([]component.Component{comp})
+								if err != nil {
+									subErrCh <- err
+								}
+							} else {
+								// got back to healthy after kill
+								subErrCh <- nil
+							}
+						} else if unit.State == client.UnitStateStarting || unit.State == client.UnitStateStopped {
+							// acceptable
+						} else {
+							// unknown state that should not have occurred
+							subErrCh <- fmt.Errorf("unit reported unexpected state: %v", unit.State)
+						}
+					} else {
+						subErrCh <- errors.New("unit missing: fake-input")
+					}
+				}
+			}
+		}
+	}()
+
+	defer drainErrChan(errCh)
+	defer drainErrChan(subErrCh)
+
+	startTimer := time.NewTimer(100 * time.Millisecond)
+	defer startTimer.Stop()
+	select {
+	case <-startTimer.C:
+		err = m.Update([]component.Component{comp})
+		require.NoError(t, err)
+	case err := <-errCh:
+		t.Fatalf("failed early: %s", err)
+	}
+
+	endTimer := time.NewTimer(1 * time.Minute)
+	defer endTimer.Stop()
+LOOP:
+	for {
+		select {
+		case <-endTimer.C:
+			t.Fatalf("timed out after 30 seconds")
+		case err := <-errCh:
+			require.NoError(t, err)
+		case err := <-subErrCh:
+			require.NoError(t, err)
+			break LOOP
+		}
+	}
+
+	subCancel()
+	cancel()
+
+	err = <-errCh
+	require.NoError(t, err)
+}
+
+func TestManager_FakeInput_KeepsRestarting(t *testing.T) {
+	testPaths(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ai, _ := info.NewAgentInfo(true)
+	m, err := NewManager(newDebugLogger(t), "localhost:0", ai, apmtest.DiscardTracer, newTestMonitoringMgr(), configuration.DefaultGRPCConfig())
+	require.NoError(t, err)
+	errCh := make(chan error)
+	go func() {
+		err := m.Run(ctx)
+		if errors.Is(err, context.Canceled) {
+			err = nil
+		}
+		errCh <- err
+	}()
+
+	waitCtx, waitCancel := context.WithTimeout(ctx, 1*time.Second)
+	defer waitCancel()
+	if err := m.WaitForReady(waitCtx); err != nil {
+		require.NoError(t, err)
+	}
+
+	// adjust input spec to allow restart
+	cmdSpec := *fakeInputSpec.Command
+	cmdSpec.RestartMonitoringPeriod = 1 * time.Second
+	cmdSpec.MaxRestartsPerPeriod = 10
+	inputSpec := fakeInputSpec
+	inputSpec.Command = &cmdSpec
+
+	binaryPath := testBinary(t, "component")
+	comp := component.Component{
+		ID: "fake-default",
+		InputSpec: &component.InputRuntimeSpec{
+			InputType:  "fake",
+			BinaryName: "",
+			BinaryPath: binaryPath,
+			Spec:       inputSpec,
+		},
+		Units: []component.Unit{
+			{
+				ID:       "fake-input",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy",
+					"killer":  true,
```

Comment:
Either add a comment that this causes the component to randomly die, or rename the config option to obviously state this (like `randomlySelfTerminate` or something). I had to go look a the fake component implementation to remember what this is doing, and it will be harder to figure out without looking with the paired change there in the same diff.

Reply:
It's not random it is on an interval. I will work on a better name.